File: Realize/6Week/PracticeGraph4.cpp
// Practice3
// 인접 리스트 그래프의 DFS, BFS
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

// Node class for adjacency list
class Node {
public:
    int id;
    Node* next;
    Node(int id, Node* next = nullptr) : id(id), next(next) {}
};

// MyGraphList base class
class MyGraphList {
public:
    vector<char> vertices;
    vector<Node*> adjList;
    int elemCnt;

    MyGraphList(int size) {
        adjList.resize(size, nullptr);
        elemCnt = 0;
    }

    void addVertex(char data) {
        vertices.push_back(data);
        elemCnt++;
    }

    void addEdge(int x, int y) {
        adjList[x] = new Node(y, adjList[x]);
        adjList[y] = new Node(x, adjList[y]);  // Undirected edge
    }

    void printAdjacentList() {
        for (int i = 0; i < elemCnt; i++) {
            cout << vertices[i] << ": ";
            Node* cur = adjList[i];
            while (cur != nullptr) {
                cout << vertices[cur->id] << " ";
                cur = cur->next;
            }
            cout << endl;
        }
    }
};

// MyGraphList2 class that extends MyGraphList and adds DFS and BFS
class MyGraphList2 : public MyGraphList {
public:
    MyGraphList2(int size) : MyGraphList(size) {}

    // Depth First Search (DFS)
    void dfs(int id) {
        vector<bool> visited(elemCnt, false);
        stack<int> stack;

        stack.push(id);
        visited[id] = true;

        while (!stack.empty()) {
            int curID = stack.top();
            stack.pop();
            cout << vertices[curID] << " ";

            Node* cur = adjList[curID];
            while (cur != nullptr) {
                if (!visited[cur->id]) {
                    stack.push(cur->id);
                    visited[cur->id] = true;
                }
                cur = cur->next;
            }
        }
        cout << endl;
    }

    // Breadth First Search (BFS)
    void bfs(int id) {
        vector<bool> visited(elemCnt, false);
        queue<int> q;

        q.push(id);
        visited[id] = true;

        while (!q.empty()) {
            int curID = q.front();
            q.pop();
            cout << vertices[curID] << " ";

            Node* cur = adjList[curID];
            while (cur != nullptr) {
                if (!visited[cur->id]) {
                    q.push(cur->id);
                    visited[cur->id] = true;
                }
                cur = cur->next;
            }
        }
        cout << endl;
    }
};

int main() {
    // Test code
    MyGraphList2 graph(7);
    graph.addVertex('A');  // 0
    graph.addVertex('B');  // 1
    graph.addVertex('C');  // 2
    graph.addVertex('D');  // 3
    graph.addVertex('E');  // 4
    graph.addVertex('F');  // 5
    graph.addVertex('G');  // 6

    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(0, 3);
    graph.addEdge(1, 4);
    graph.addEdge(2, 5);
    graph.addEdge(3, 4);
    graph.addEdge(3, 5);
    graph.addEdge(4, 6);
    graph.addEdge(5, 6);

    graph.printAdjacentList();

    cout << "DFS: ";
    graph.dfs(0);

    cout << "BFS: ";
    graph.bfs(0);

    return 0;
}

File: Realize/6Week/PracticeGraph4.py
# Practice3
# 인접 리스트 그래프의 DFS, BFS
class Node:
    def __init__(self, id, next_node=None):
        self.id = id
        self.next = next_node

class MyGraphList:
    def __init__(self, size):
        self.vertices = []
        self.adjList = [None] * size
        self.elemCnt = 0

    def addVertex(self, data):
        self.vertices.append(data)
        self.elemCnt += 1

    def addEdge(self, x, y):
        self.adjList[x] = Node(y, self.adjList[x])
        self.adjList[y] = Node(x, self.adjList[y])  # Undirected edge

    def printAdjacentList(self):
        for i in range(self.elemCnt):
            print(f"{self.vertices[i]}:", end=" ")
            cur = self.adjList[i]
            while cur:
                print(self.vertices[cur.id], end=" ")
                cur = cur.next
            print()

class MyGraphList2(MyGraphList):
    def __init__(self, size):
        super().__init__(size)

    def dfs(self, id):
        visited = [False] * self.elemCnt
        stack = [id]
        visited[id] = True

        while stack:
            curID = stack.pop()
            print(self.vertices[curID], end=" ")

            cur = self.adjList[curID]
            while cur:
                if not visited[cur.id]:
                    stack.append(cur.id)
                    visited[cur.id] = True
                cur = cur.next
        print()

    def bfs(self, id):
        visited = [False] * self.elemCnt
        queue = [id]
        visited[id] = True

        while queue:
            curID = queue.pop(0)
            print(self.vertices[curID], end=" ")

            cur = self.adjList[curID]
            while cur:
                if not visited[cur.id]:
                    queue.append(cur.id)
                    visited[cur.id] = True
                cur = cur.next
        print()

if __name__ == "__main__":
    graph = MyGraphList2(7)
    graph.addVertex('A')  # 0
    graph.addVertex('B')  # 1
    graph.addVertex('C')  # 2
    graph.addVertex('D')  # 3
    graph.addVertex('E')  # 4
    graph.addVertex('F')  # 5
    graph.addVertex('G')  # 6

    graph.addEdge(0, 1)
    graph.addEdge(0, 2)
    graph.addEdge(0, 3)
    graph.addEdge(1, 4)
    graph.addEdge(2, 5)
    graph.addEdge(3, 4)
    graph.addEdge(3, 5)
    graph.addEdge(4, 6)
    graph.addEdge(5, 6)

    graph.printAdjacentList()

    print("DFS:", end=" ")
    graph.dfs(0)

    print("BFS:", end=" ")
    graph.bfs(0)
